File: webview-bridge-sdk/ios/Sources/Modules/ClipboardModule.swift
/**
 * Clipboard 模块 - 剪贴板操作
 *
 * 提供剪贴板读写功能，支持文本、图片等多种格式
 */

import Foundation
import UIKit
import MobileCoreServices
import UniformTypeIdentifiers

// MARK: - 剪贴板内容类型

/// 剪贴板内容类型
public enum ClipboardContentType: String {
    case text
    case html
    case url
    case image
}

// MARK: - Clipboard 模块

public class ClipboardModule: BridgeModule {
    
    public let moduleName = "Clipboard"
    public let methods = ["Read", "Write", "HasContent", "Clear", "GetAvailableTypes"]
    
    private weak var bridge: WebViewBridge?
    
    public init(bridge: WebViewBridge) {
        self.bridge = bridge
    }
    
    public func handleRequest(
        method: String,
        params: [String: AnyCodable],
        callback: @escaping (Result<Any?, BridgeError>) -> Void
    ) {
        switch method {
        case "Read":
            read(params: params, callback: callback)
        case "Write":
            write(params: params, callback: callback)
        case "HasContent":
            hasContent(params: params, callback: callback)
        case "Clear":
            clear(callback: callback)
        case "GetAvailableTypes":
            getAvailableTypes(callback: callback)
        default:
            callback(.failure(BridgeError.methodNotFound("\(moduleName).\(method)")))
        }
    }
    
    // MARK: - Read
    
    private func read(
        params: [String: AnyCodable],
        callback: @escaping (Result<Any?, BridgeError>) -> Void
    ) {
        let typeString = params.getString("type") ?? "text"
        let type = ClipboardContentType(rawValue: typeString) ?? .text
        
        let pasteboard = UIPasteboard.general
        
        switch type {
        case .text:
            if let text = pasteboard.string {
                callback(.success([
                    "type": "text",
                    "content": text
                ]))
            } else {
                callback(.success([
                    "type": "text",
                    "content": NSNull()
                ]))
            }
            
        case .html:
            // 尝试读取 HTML 内容
            if #available(iOS 14.0, *) {
                if let data = pasteboard.data(forPasteboardType: UTType.html.identifier),
                   let html = String(data: data, encoding: .utf8) {
                    callback(.success([
                        "type": "html",
                        "content": html
                    ]))
                } else {
                    callback(.success([
                        "type": "html",
                        "content": NSNull()
                    ]))
                }
            } else {
                if let data = pasteboard.data(forPasteboardType: kUTTypeHTML as String),
                   let html = String(data: data, encoding: .utf8) {
                    callback(.success([
                        "type": "html",
                        "content": html
                    ]))
                } else {
                    callback(.success([
                        "type": "html",
                        "content": NSNull()
                    ]))
                }
            }
            
        case .url:
            if let url = pasteboard.url {
                callback(.success([
                    "type": "url",
                    "content": url.absoluteString
                ]))
            } else if let urlString = pasteboard.string,
                      URL(string: urlString) != nil {
                callback(.success([
                    "type": "url",
                    "content": urlString
                ]))
            } else {
                callback(.success([
                    "type": "url",
                    "content": NSNull()
                ]))
            }
            
        case .image:
            if let image = pasteboard.image,
               let base64 = image.pngData()?.base64EncodedString() {
                callback(.success([
                    "type": "image",
                    "content": "data:image/png;base64,\(base64)"
                ]))
            } else {
                callback(.success([
                    "type": "image",
                    "content": NSNull()
                ]))
            }
        }
    }
    
    // MARK: - Write
    
    private func write(
        params: [String: AnyCodable],
        callback: @escaping (Result<Any?, BridgeError>) -> Void
    ) {
        let typeString = params.getString("type") ?? "text"
        let type = ClipboardContentType(rawValue: typeString) ?? .text
        
        guard let content = params.getString("content") else {
            callback(.failure(BridgeError.invalidParams("content")))
            return
        }
        
        let pasteboard = UIPasteboard.general
        
        switch type {
        case .text:
            pasteboard.string = content
            callback(.success(nil))
            
        case .html:
            // 写入 HTML 内容
            if #available(iOS 14.0, *) {
                if let data = content.data(using: .utf8) {
                    pasteboard.setData(data, forPasteboardType: UTType.html.identifier)
                    // 同时写入纯文本版本
                    pasteboard.string = content.stripHTML()
                }
            } else {
                if let data = content.data(using: .utf8) {
                    pasteboard.setData(data, forPasteboardType: kUTTypeHTML as String)
                    pasteboard.string = content.stripHTML()
                }
            }
            callback(.success(nil))
            
        case .url:
            if let url = URL(string: content) {
                pasteboard.url = url
                callback(.success(nil))
            } else {
                callback(.failure(BridgeError.invalidParams("无效的 URL")))
            }
            
        case .image:
            // 支持 base64 或 URL
            if content.hasPrefix("data:image") {
                // Base64 图片
                if let base64Data = extractBase64Data(from: content),
                   let data = Data(base64Encoded: base64Data),
                   let image = UIImage(data: data) {
                    pasteboard.image = image
                    callback(.success(nil))
                } else {
                    callback(.failure(BridgeError.invalidParams("无效的 Base64 图片")))
                }
            } else {
                callback(.failure(BridgeError.invalidParams("图片格式不支持")))
            }
        }
    }
    
    // MARK: - HasContent
    
    private func hasContent(
        params: [String: AnyCodable],
        callback: @escaping (Result<Any?, BridgeError>) -> Void
    ) {
        let typeString = params.getString("type") ?? "text"
        let type = ClipboardContentType(rawValue: typeString) ?? .text
        
        let pasteboard = UIPasteboard.general
        var hasContent = false
        
        switch type {
        case .text:
            hasContent = pasteboard.hasStrings
        case .html:
            if #available(iOS 14.0, *) {
                hasContent = pasteboard.contains(pasteboardTypes: [UTType.html.identifier])
            } else {
                hasContent = pasteboard.contains(pasteboardTypes: [kUTTypeHTML as String])
            }
        case .url:
            hasContent = pasteboard.hasURLs
        case .image:
            hasContent = pasteboard.hasImages
        }
        
        callback(.success([
            "type": typeString,
            "hasContent": hasContent
        ]))
    }
    
    // MARK: - Clear
    
    private func clear(callback: @escaping (Result<Any?, BridgeError>) -> Void) {
        UIPasteboard.general.items = []
        callback(.success(nil))
    }
    
    // MARK: - GetAvailableTypes
    
    private func getAvailableTypes(callback: @escaping (Result<Any?, BridgeError>) -> Void) {
        let pasteboard = UIPasteboard.general
        var types: [String] = []
        
        if pasteboard.hasStrings {
            types.append("text")
        }
        if pasteboard.hasURLs {
            types.append("url")
        }
        if pasteboard.hasImages {
            types.append("image")
        }
        
        if #available(iOS 14.0, *) {
            if pasteboard.contains(pasteboardTypes: [UTType.html.identifier]) {
                types.append("html")
            }
        } else {
            if pasteboard.contains(pasteboardTypes: [kUTTypeHTML as String]) {
                types.append("html")
            }
        }
        
        callback(.success([
            "types": types,
            "isEmpty": types.isEmpty
        ]))
    }
    
    // MARK: - 辅助方法
    
    private func extractBase64Data(from dataURL: String) -> String? {
        // 格式: data:image/png;base64,XXXX
        guard let commaIndex = dataURL.firstIndex(of: ",") else {
            return nil
        }
        return String(dataURL[dataURL.index(after: commaIndex)...])
    }
}

// MARK: - String 扩展

private extension String {
    /// 移除 HTML 标签
    func stripHTML() -> String {
        guard let data = self.data(using: .utf8) else { return self }
        
        let options: [NSAttributedString.DocumentReadingOptionKey: Any] = [
            .documentType: NSAttributedString.DocumentType.html,
            .characterEncoding: String.Encoding.utf8.rawValue
        ]
        
        if let attributedString = try? NSAttributedString(data: data, options: options, documentAttributes: nil) {
            return attributedString.string
        }
        
        return self
    }
}

File: android/src/main/java/com/aspect/webviewbridge/modules/ClipboardModule.kt
/**
 * Clipboard 模块 - 剪贴板操作
 *
 * 提供剪贴板读写功能，支持文本、HTML 等格式
 */

package com.aspect.webviewbridge.modules

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.os.Build
import android.util.Base64
import com.aspect.webviewbridge.protocol.*

/**
 * 剪贴板内容类型
 */
enum class ClipboardContentType {
    TEXT,
    HTML,
    URL,
    IMAGE
}

/**
 * Clipboard 模块
 */
class ClipboardModule(
    private val context: Context,
    private val bridgeContext: BridgeModuleContext
) : BridgeModule {

    override val moduleName = "Clipboard"

    override val methods = listOf(
        "Read",
        "Write",
        "HasContent",
        "Clear",
        "GetAvailableTypes"
    )

    private val clipboardManager: ClipboardManager by lazy {
        context.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
    }

    override fun handleRequest(
        method: String,
        request: BridgeRequest,
        callback: (Result<Any?>) -> Unit
    ) {
        when (method) {
            "Read" -> read(request, callback)
            "Write" -> write(request, callback)
            "HasContent" -> hasContent(request, callback)
            "Clear" -> clear(callback)
            "GetAvailableTypes" -> getAvailableTypes(callback)
            else -> callback(Result.failure(BridgeError.methodNotFound("$moduleName.$method")))
        }
    }

    // MARK: - Read

    private fun read(
        request: BridgeRequest,
        callback: (Result<Any?>) -> Unit
    ) {
        val typeString = request.getString("type") ?: "text"
        val type = try {
            ClipboardContentType.valueOf(typeString.uppercase())
        } catch (e: Exception) {
            ClipboardContentType.TEXT
        }

        val clipData = clipboardManager.primaryClip

        if (clipData == null || clipData.itemCount == 0) {
            callback(
                Result.success(
                    mapOf(
                        "type" to typeString,
                        "content" to null
                    )
                )
            )
            return
        }

        val item = clipData.getItemAt(0)

        when (type) {
            ClipboardContentType.TEXT -> {
                val text = item.text?.toString()
                callback(
                    Result.success(
                        mapOf(
                            "type" to "text",
                            "content" to text
                        )
                    )
                )
            }

            ClipboardContentType.HTML -> {
                val html = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                    item.htmlText ?: item.text?.toString()
                } else {
                    item.text?.toString()
                }
                callback(
                    Result.success(
                        mapOf(
                            "type" to "html",
                            "content" to html
                        )
                    )
                )
            }

            ClipboardContentType.URL -> {
                val uri = item.uri?.toString() ?: item.text?.toString()
                callback(
                    Result.success(
                        mapOf(
                            "type" to "url",
                            "content" to uri
                        )
                    )
                )
            }

            ClipboardContentType.IMAGE -> {
                // Android 剪贴板图片处理较复杂，需要 ContentResolver
                val uri = item.uri
                if (uri != null) {
                    try {
                        context.contentResolver.openInputStream(uri)?.use { stream ->
                            val bytes = stream.readBytes()
                            val base64 = Base64.encodeToString(bytes, Base64.NO_WRAP)
                            callback(
                                Result.success(
                                    mapOf(
                                        "type" to "image",
                                        "content" to "data:image/png;base64,$base64"
                                    )
                                )
                            )
                        } ?: run {
                            callback(
                                Result.success(
                                    mapOf(
                                        "type" to "image",
                                        "content" to null
                                    )
                                )
                            )
                        }
                    } catch (e: Exception) {
                        callback(
                            Result.success(
                                mapOf(
                                    "type" to "image",
                                    "content" to null
                                )
                            )
                        )
                    }
                } else {
                    callback(
                        Result.success(
                            mapOf(
                                "type" to "image",
                                "content" to null
                            )
                        )
                    )
                }
            }
        }
    }

    // MARK: - Write

    private fun write(
        request: BridgeRequest,
        callback: (Result<Any?>) -> Unit
    ) {
        val typeString = request.getString("type") ?: "text"
        val content = request.getString("content")

        if (content == null) {
            callback(Result.failure(BridgeError.invalidParams("content")))
            return
        }

        val type = try {
            ClipboardContentType.valueOf(typeString.uppercase())
        } catch (e: Exception) {
            ClipboardContentType.TEXT
        }

        try {
            val clip = when (type) {
                ClipboardContentType.TEXT -> {
                    ClipData.newPlainText("text", content)
                }

                ClipboardContentType.HTML -> {
                    if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN) {
                        ClipData.newHtmlText("html", stripHtml(content), content)
                    } else {
                        ClipData.newPlainText("html", content)
                    }
                }

                ClipboardContentType.URL -> {
                    ClipData.newPlainText("url", content)
                }

                ClipboardContentType.IMAGE -> {
                    // 图片写入需要 ContentProvider，这里简化处理
                    callback(Result.failure(BridgeError.notSupported("图片写入暂不支持")))
                    return
                }
            }

            clipboardManager.setPrimaryClip(clip)
            callback(Result.success(null))
        } catch (e: Exception) {
            callback(Result.failure(BridgeError.internalError(e.message)))
        }
    }

    // MARK: - HasContent

    private fun hasContent(
        request: BridgeRequest,
        callback: (Result<Any?>) -> Unit
    ) {
        val typeString = request.getString("type") ?: "text"

        val hasContent = clipboardManager.hasPrimaryClip() &&
                clipboardManager.primaryClip?.itemCount ?: 0 > 0

        callback(
            Result.success(
                mapOf(
                    "type" to typeString,
                    "hasContent" to hasContent
                )
            )
        )
    }

    // MARK: - Clear

    private fun clear(callback: (Result<Any?>) -> Unit) {
        try {
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                clipboardManager.clearPrimaryClip()
            } else {
                clipboardManager.setPrimaryClip(ClipData.newPlainText("", ""))
            }
            callback(Result.success(null))
        } catch (e: Exception) {
            callback(Result.failure(BridgeError.internalError(e.message)))
        }
    }

    // MARK: - GetAvailableTypes

    private fun getAvailableTypes(callback: (Result<Any?>) -> Unit) {
        val types = mutableListOf<String>()

        val clipData = clipboardManager.primaryClip
        if (clipData != null && clipData.itemCount > 0) {
            val description = clipData.description

            // 检查 MIME 类型
            for (i in 0 until description.mimeTypeCount) {
                when {
                    description.getMimeType(i).startsWith("text/html") -> {
                        if (!types.contains("html")) types.add("html")
                    }

                    description.getMimeType(i).startsWith("text/") -> {
                        if (!types.contains("text")) types.add("text")
                    }

                    description.getMimeType(i).startsWith("image/") -> {
                        if (!types.contains("image")) types.add("image")
                    }
                }
            }

            // 检查 URI
            val item = clipData.getItemAt(0)
            if (item.uri != null && !types.contains("url")) {
                types.add("url")
            }

            // 确保至少有 text
            if (item.text != null && !types.contains("text")) {
                types.add("text")
            }
        }

        callback(
            Result.success(
                mapOf(
                    "types" to types,
                    "isEmpty" to types.isEmpty()
                )
            )
        )
    }

    // MARK: - 辅助方法

    private fun stripHtml(html: String): String {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            android.text.Html.fromHtml(html, android.text.Html.FROM_HTML_MODE_LEGACY).toString()
        } else {
            @Suppress("DEPRECATION")
            android.text.Html.fromHtml(html).toString()
        }
    }
}
